sql/api/src/main/scala/org/apache/spark/sql/types/UpCastRule.scala

```diff
@@ -40,8 +40,8 @@ private[sql] object UpCastRule {
     case (DateType, TimestampNTZType) => true
     case (TimestampNTZType, TimestampType) => true
     case (TimestampType, TimestampNTZType) => true
-    case (_: AtomicType, StringType) => true
-    case (_: CalendarIntervalType, StringType) => true
+    case (_: AtomicType, _: StringType) => true
```

Comment:
can we add some tests for upcast [here](https://github.com/apache/spark/blob/400a8d3797bdcc9183576e66e84163e4dc00a662/sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuiteBase.scala#L659)?

Reply:
Also please make sure that this does not allow casting from int to string directly by calling cast. In general by standard we should do `cast (value as string) collate collation`, but it is a good catch that we need to allow this transition in other operations.